src/dtx/dtx_srv.c

```diff
@@ -14,13 +14,81 @@
 #include <daos_srv/container.h>
 #include <daos_srv/vos.h>
 #include <daos_srv/dtx_srv.h>
+#include <gurt/telemetry_common.h>
+#include <gurt/telemetry_producer.h>
 #include "dtx_internal.h"
 
 #define DTX_YIELD_CYCLE		(DTX_THRESHOLD_COUNT >> 3)
 
+static inline char *
+dtx_opc_to_str(crt_opcode_t opc)
+{
+	switch (opc) {
+#define X(a, b, c, d, e, f) case a: return f;
+		DTX_PROTO_SRV_RPC_LIST
+#undef X
+	}
+	return "dtx_unknown";
+}
+
+struct dtx_tls {
+	struct d_tm_node_t	*ot_op_total[DTX_PROTO_SRV_RPC_COUNT];
+};
+
+static void *
+dtx_tls_init(int xs_id, int tgt_id)
+{
+	struct dtx_tls	*tls;
+	char		*path;
+	uint32_t	 opc;
+	int		 rc;
+
+	D_ALLOC_PTR(tls);
+	if (tls == NULL)
+		return NULL;
+
+	/** Skip sensor setup on system xstreams */
+	if (tgt_id < 0)
+		return tls;
+
+	/** Register different per-opcode sensors */
+	for (opc = 0; opc < DTX_PROTO_SRV_RPC_COUNT; opc++) {
+		D_ASPRINTF(path, "io/%u/ops/%s/total_cnt", tgt_id,
+			   dtx_opc_to_str(opc));
+		rc = d_tm_add_metric(&tls->ot_op_total[opc], path, D_TM_COUNTER,
+				     "total number of processed DTX RPCs", "");
+		if (rc != 0)
```

Comment:
There will be a change from D_TM_SUCCESS to DER_SUCCESS.  I suggest using either constant right now in place of testing rc against 0.  The warning could be made more helpful if you add the dtx_opc_to_str(opc) to the warning message so that it indicates which specific metric wasn't added.  If this were to fail, it will have more helpful data that way.  Your initialization and usage look good to me.

Reply:
OK, I will refresh the patch.